sycl/doc/design/DeviceIf.md

```diff
@@ -338,4 +338,1011 @@ in that compilation phase.
+
+  ```
+  call void @call_if_on_device_conditionally_GUID_N(
+              @CallableXXX, %callablethis, N1, N2, ...)
+  ```
+
+  * Using the device configuration file, determine if the _Condition
+    Expression_ is true for this target.
+
+  * If the condition is true, change the call site into a direct call to the
+    _Conditional Action_ function like so:
+
+    ```
+    call void @CallableXXX(%callablethis)
+    ```
+
+  * If the condition is false, remove the call site entirely.
+
+* Because we did not previously propagate the `!sycl_used_aspects` from the
+  _Conditional Action_ functions up to their containing kernels, this must be
+  done now.  For each call to a _Conditional Action_ function, add its
+  `!sycl_used_aspects` set to each kernel that is a predecessor in the static
+  call tree.
+
+After this processing is complete, the device code is split into images in the
+normal way, as described in the [Optional Device Features][5] design.
+
+[5]: <./OptionalDeviceFeatures.md#changes-to-the-device-code-split-algorithm>
+
+### New property sets
+[sec-prop-sets]: <#new-property-sets>
+
+When the `sycl-post-link` tool runs in non-AOT mode, it needs to communicate
+information to the DPC++ runtime about the _Add On Images_ that contain the
+_Conditional Action_ functions.  We do this by adding two new property sets.
+
+The _Main Image_ contains a property set named "SYCL/add on images" which
+contains one property for each of its _Add On Images_.  The name of each
+property is a unique identifier for the _Add On Image_, which by convention is
+just an integer in string form (e.g "1").  The value of the property has type
+`PI_PROPERTY_TYPE_BYTE_ARRAY` containing a series of `uint32` values `N1`,
+`N2`, etc. telling the _Condition Expression_ for the _Conditional Actions_ in
+this _Add On Image_.  The property's size (which is always divisible by 4)
+tells the number of `uint32` integers in this property.
+
+```
+[SYCL/add on images]
+"1": N1, N2, N3, ...
+"2": N1, N2, N3, ...
+```
+
+Each _Add On Image_ contains a property set named "SYCL/add on image" which
+contains just one property.  The name of the property is the unique identifier
+for this _Add On Image_ (e.g. "1").  The value of the property has type
+`PI_PROPERTY_TYPE_UINT32`.  The value is `1` if the _Add On Image_ contains the
+real function definitions, and it is `0` if the _Add On Image_ contains the
+stub definitions.
+
+```
+[SYCL/add on image]
+"1": 1
+```
+
+### Changes to the DPC++ runtime
+
+The DPC++ runtime requires changes to identify and dynamically link the _Add On
+Images_ when necessary.
+
+When the application submits a kernel to a device, the runtime must already
+find the device images from shared libraries that define exported device
+functions as described in [Device Code Dynamic Linking][6].  This algorithm is
+extended to look also for _Add On Images_.
+
+If the _Main Image_ contains the "SYCL/add on images" property set, the runtime
```

Comment:
How would this handle nested add-on images, would these all be contained in the same property set, does the linking order here matter?

Reply:
Nested add-on images are handled in the last bullet point below that starts "The selected _Add On Image_ may also contain a "SYCL/add on images" property set".

The algorithm in `sycl-post-link` breaks nested `if_device_has` calls into separate add-on images.  Each of those add-on images contains its own "SYCL/device requirements" property set that defines the aspects it uses.

The link order doesn't matter.